devtools/pip-tools/pr_tool/backport.py

```diff
@@ -0,0 +1,317 @@
+    remote_branch = local_branch.tracking_branch()
+    if remote_branch.remote_name != git_remote.name:
+        logging.error(
+            "%s does not track a branch on remote %s", branch_name, git_remote
+        )
+        return None
+    logging.info("Remote branch is '%s'", remote_branch.remote_head)
+    return remote_branch.remote_head
+
+
+def get_pr_metadata(*, repo):
+    cfg = repo.head.reference.config_reader()
+    if not cfg.has_option("bareos-original-pr") or not cfg.has_option(
+        "bareos-backport-into"
+    ):
+        raise UnmanagedBranchException(
+            f"Branch {repo.head.reference.name} is not a managed backport branch."
+        )
+    pr = cfg.get_value("bareos-original-pr")
+    into = cfg.get_value("bareos-backport-into")
+    return (pr, into)
+
+
+def cherry_pick(*, repo, upstream, select_all=False, allow_reset=False):
+    try:
+        original_pr, base_branch_name = get_pr_metadata(repo=repo)
+    except UnmanagedBranchException as e:
+        logging.critical(e)
+        return False
+
+    upstream.fetch([base_branch_name, f"refs/pull/{original_pr}/head"])
+
+    base_commit = upstream.refs[base_branch_name]
+    if repo.head.commit != base_commit:
+        if allow_reset:
+            logging.info("Resetting HEAD to %s", base_commit)
+            repo.head.reset(base_commit, working_tree=True, index=True)
+        else:
+            logging.critical(
+                "Your HEAD does not match %s. Either rerun with --reset or reset manually first.",
+                base_commit,
+            )
+            return False
+
+    pr_data = Gh().pr.view(str(original_pr), json=["title", "commits"])
+    return cherry_pick_impl(
+        repo=repo,
+        original_pr=original_pr,
+        commits=pr_data["commits"],
+        title=pr_data["title"],
+        select_all=select_all,
+    )
+
+
+def generate_commit_lines(original_pr, title, commits):
+    lines = [
+        f"# Backport of #{original_pr}: {title}\n",
+        "# List all commits to cherry-pick in order, remove commits you do not want.\n",
+    ]
+    for c in commits:
+        if c["messageHeadline"] == "Update CHANGELOG.md":
+            lines.append(f"# {c['oid'][0:9]} {c['messageHeadline']}\n")
+        else:
+            lines.append(f"{c['oid'][0:9]} {c['messageHeadline']}\n")
+    return lines
+
+
+def parse_commit_lines(lines):
+    commit_lines = [
+        l.partition(" ") for l in lines if not l.startswith("#") and l.strip() != ""
+    ]
+    todo = []
+    for commit, space, descr in commit_lines:
+        if space != " " or not descr.endswith("\n"):
+            logging.critical("Cannot parse '%s%s%s', abort.", commit, space, descr)
+            return False
+        todo.append(commit)
+    return todo
+
+
+def cherry_pick_impl(*, repo, original_pr, commits, title, select_all=False):
+    lines = generate_commit_lines(original_pr, title, commits)
+
+    if not select_all:
+        with NamedTemporaryFile("r+") as fp:
```

Comment:
It will exist (`NamedTemporaryFile` creates the temporary file for us) and it cannot have existed already. `NamedTemporaryFile`'s main purpose is to deal with exactly these kinds of edge-cases)

Reply:
as `w+b` is the default I'll just go with `w+` even if I don't understand why :)